File: sdk/ts/test/login.test.ts
import { test } from "node:test";
import assert from "node:assert/strict";

import { getPublicKey, sign } from "@bitspark/archon";
import {
  loginBinding,
  proveLogin,
  verifyLogin,
  proveCollect,
  verifyCollect,
  provePossession,
  possessionMessageBytes,
  LOGIN_DOMAIN,
  LOGIN_ROLE_LOGIN,
  LOGIN_ROLE_COLLECT,
  LOGIN_MAX_FIELD_SIZE,
  deriveAudience,
  type LoginRequest,
} from "../src/index.js";

const seedP = new Uint8Array(32).fill(0x11);
const seedK = new Uint8Array(32).fill(0x22);
const audience = "https://dawn.example/api";
const utf8 = new TextEncoder();

function request(): LoginRequest {
  return {
    id: utf8.encode("req-1"),
    nonce: new Uint8Array(16).fill(0xaa),
    browser: getPublicKey(seedK),
    scope: ["read:projects", "read:campaigns"],
    validFor: 28800,
  };
}

test("binding layout is the pinned one; roles differ in the first byte", () => {
  const req = request();
  const b = loginBinding(LOGIN_ROLE_LOGIN, audience, req);
  const want = [
    0x01, 0x00, audience.length, ...utf8.encode(audience), ...req.browser,
    0x00, 0x05, ...utf8.encode("req-1"),
    0x00, 0x02, 0x00, 0x0d, ...utf8.encode("read:projects"), 0x00, 0x0e, ...utf8.encode("read:campaigns"),
    0x00, 0x00, 0x70, 0x80,
  ];
  assert.deepEqual([...b], want);
  const c = loginBinding(LOGIN_ROLE_COLLECT, audience, req);
  assert.equal(c[0], LOGIN_ROLE_COLLECT);
  assert.deepEqual([...c.subarray(1)], [...b.subarray(1)]);
});

test("login proof round-trips and every bound field binds", () => {
  const req = request();
  const sig = proveLogin(seedP, audience, req);
  const pubP = getPublicKey(seedP);
  assert.equal(verifyLogin(pubP, audience, req, sig), true);
  assert.equal(verifyLogin(pubP, "https://evil.example/api", req, sig), false);
  assert.equal(verifyLogin(pubP, audience, { ...req, browser: pubP }, sig), false);
  assert.equal(verifyLogin(pubP, audience, { ...req, id: utf8.encode("req-2") }, sig), false);
  assert.equal(verifyLogin(pubP, audience, { ...req, nonce: new Uint8Array(16).fill(0xab) }, sig), false);
  assert.equal(verifyLogin(pubP, audience, { ...req, scope: [...req.scope].reverse() }, sig), false);
  assert.equal(verifyLogin(pubP, audience, { ...req, scope: [...req.scope, "admin"] }, sig), false);
  assert.equal(verifyLogin(pubP, audience, { ...req, validFor: 28801 }, sig), false);
  assert.equal(verifyLogin(getPublicKey(seedK), audience, req, sig), false);
  const binding = loginBinding(LOGIN_ROLE_LOGIN, audience, req);
  const raw = sign(possessionMessageBytes(req.nonce, binding), seedP);
  assert.equal(verifyLogin(pubP, audience, req, raw), false, "a raw signature is not a proof");
  const other = provePossession(seedP, "archon-login/2", req.nonce, binding);
  assert.equal(verifyLogin(pubP, audience, req, other), false, "another domain is not a proof");
  const collect = proveCollect(seedK, audience, req);
  assert.equal(verifyLogin(getPublicKey(seedK), audience, req, collect), false, "the collect role is not a login proof");
  assert.equal(LOGIN_DOMAIN, "archon-login/1");
});

test("collect proof: only from the key the request names", () => {
  const req = request();
  const sig = proveCollect(seedK, audience, req);
  assert.equal(verifyCollect(audience, req, sig), true);
  assert.equal(verifyCollect("https://evil.example/api", req, sig), false);
  assert.equal(verifyCollect(audience, req, proveLogin(seedK, audience, req)), false);
  assert.throws(() => proveCollect(seedP, audience, req));
});

test("refusals", () => {
  const cases: Array<[string, string, Partial<LoginRequest>]> = [
    ["empty audience", "", {}],
    ["control char in audience", "https://dawn.example/api\n", {}],
    ["browser wrong size", audience, { browser: new Uint8Array(31) }],
    ["empty id", audience, { id: new Uint8Array(0) }],
    ["empty scope entry", audience, { scope: [""] }],
    ["control char in scope", audience, { scope: ["read:\u0007projects"] }],
    ["DEL in scope", audience, { scope: ["read:\u007fprojects"] }],
    ["lone surrogate in scope", audience, { scope: ["read:\ud800projects"] }],
    ["zero validity", audience, { validFor: 0 }],
    ["validity over u32", audience, { validFor: 0x1_0000_0000 }],
    ["fractional validity", audience, { validFor: 1.5 }],
    ["oversized id", audience, { id: new Uint8Array(LOGIN_MAX_FIELD_SIZE + 1) }],
    ["binding over the possession bound", audience, { scope: ["a".repeat(LOGIN_MAX_FIELD_SIZE), "b".repeat(LOGIN_MAX_FIELD_SIZE)] }],
  ];
  for (const [name, aud, patch] of cases) {
    const req = { ...request(), ...patch };
    assert.throws(() => loginBinding(LOGIN_ROLE_LOGIN, aud, req), name);
    assert.throws(() => proveLogin(seedP, aud, req), `${name}: prove`);
    assert.equal(verifyLogin(getPublicKey(seedP), aud, req, new Uint8Array(64)), false, `${name}: verify`);
  }
  assert.throws(() => loginBinding(0x03, audience, request()));
  assert.throws(() => proveLogin(seedP, audience, { ...request(), nonce: new Uint8Array(15) }));
  assert.throws(() => proveLogin(seedP.subarray(0, 31), audience, request()));
  assert.doesNotThrow(() => loginBinding(LOGIN_ROLE_LOGIN, audience, { ...request(), scope: [] }), "empty scope list is allowed");
});

test("deriveAudience: the §2.1 grammar, refused not normalised", () => {
  const id = "8f3c1d2e4b5a69780f1e2d3c4b5a6978";
  const accepts: Array<[string, string]> = [
    [`https://dawn.example/api/login/${id}`, "https://dawn.example/api"],
    [`https://dawn.example:443/api/login/${id}`, "https://dawn.example/api"],
    [`http://localhost:8080/login/${id}`, "http://localhost:8080"],
    [`HTTPS://Dawn.Example/API/login/${id}`, "https://dawn.example/API"],
    [`wss://dawn.example:80/login/${id}`, "https://dawn.example:80"],
    [`https://[2001:DB8::1]:443/login/${id}`, "https://[2001:db8::1]"],
    [`https://dawn.example/a/b%2Fc/login/${id}`, "https://dawn.example/a/b%2Fc"],
  ];
  for (const [url, want] of accepts) {
    const d = deriveAudience(url);
    assert.equal(d.audience, want, url);
    assert.equal(d.id.length, 16, url);
  }
  for (const url of [
    `dawn.example/login/${id}`,
    `ftp://dawn.example/login/${id}`,
    `https://dawn.example/api/login/${id}/`,
    `https://user:pass@dawn.example/api/login/${id}`,
    `https://dawn.example/login/${id}?x=1`,
    `https://dawn.example:0443/login/${id}`,
    "https://dawn.example/api/login/8F3C",
    "https://dawn.example/api/login/8f3",
    `https://dawn.example/a b/login/${id}`,
    `https://dawn.example/a%2/login/${id}`,
    `https://[fe80::1%25eth0]/login/${id}`,
    "",
  ]) {
    assert.throws(() => deriveAudience(url), url);
  }
});

File: sdk/go/login/login_test.go
package login

import (
	"bytes"
	"encoding/hex"
	"strings"
	"testing"

	"github.com/Bitspark/archon/core/go/crypto"
	"github.com/Bitspark/archon/sdk/go/possession"
)

var (
	seedP = bytes.Repeat([]byte{0x11}, 32)
	seedK = bytes.Repeat([]byte{0x22}, 32)
)

func request() *Request {
	return &Request{
		ID:       []byte("req-1"),
		Nonce:    bytes.Repeat([]byte{0xaa}, 16),
		Browser:  crypto.PublicKeyFromSeed(seedK),
		Scope:    []string{"read:projects", "read:campaigns"},
		ValidFor: 28800,
	}
}

const audience = "https://dawn.example/api"

func TestBindingLayout(t *testing.T) {
	req := request()
	b, err := Binding(RoleLogin, audience, req)
	if err != nil {
		t.Fatal(err)
	}
	want := []byte{0x01}
	want = append(want, 0x00, byte(len(audience)))
	want = append(want, audience...)
	want = append(want, req.Browser...)
	want = append(want, 0x00, 0x05)
	want = append(want, "req-1"...)
	want = append(want, 0x00, 0x02)
	want = append(want, 0x00, 0x0d)
	want = append(want, "read:projects"...)
	want = append(want, 0x00, 0x0e)
	want = append(want, "read:campaigns"...)
	want = append(want, 0x00, 0x00, 0x70, 0x80)
	if !bytes.Equal(b, want) {
		t.Fatalf("binding\n got %x\nwant %x", b, want)
	}
	c, _ := Binding(RoleCollect, audience, req)
	if c[0] != RoleCollect || !bytes.Equal(c[1:], b[1:]) {
		t.Fatal("collect binding differs from login binding beyond the role byte")
	}
}

func TestProveVerifyRoundTrip(t *testing.T) {
	req := request()
	sig, err := Prove(seedP, audience, req)
	if err != nil {
		t.Fatal(err)
	}
	pubP := crypto.PublicKeyFromSeed(seedP)
	if !Verify(pubP, audience, req, sig) {
		t.Fatal("login proof did not verify")
	}
	// The proof IS the possession scheme in Domain over the login binding.
	binding, _ := Binding(RoleLogin, audience, req)
	if !possession.Verify(pubP, Domain, req.Nonce, binding, sig) {
		t.Fatal("login proof is not a possession proof over the binding")
	}
	// Every bound field binds.
	tamper := []struct {
		name string
		aud  string
		mod  func(*Request)
	}{
		{"audience", "https://evil.example/api", func(*Request) {}},
		{"browser", audience, func(r *Request) { r.Browser = crypto.PublicKeyFromSeed(seedP) }},
		{"id", audience, func(r *Request) { r.ID = []byte("req-2") }},
		{"nonce", audience, func(r *Request) { r.Nonce = bytes.Repeat([]byte{0xab}, 16) }},
		{"scope order", audience, func(r *Request) { r.Scope = []string{"read:campaigns", "read:projects"} }},
		{"scope entry", audience, func(r *Request) { r.Scope = []string{"read:projects", "write:campaigns"} }},
		{"scope extra", audience, func(r *Request) { r.Scope = append(r.Scope, "admin") }},
		{"valid_for", audience, func(r *Request) { r.ValidFor = 28801 }},
	}
	for _, tc := range tamper {
		r := request()
		tc.mod(r)
		if Verify(pubP, tc.aud, r, sig) {
			t.Errorf("%s did not bind", tc.name)
		}
	}
	// Not a proof: the other key, a raw signature, another domain, the other role.
	if Verify(crypto.PublicKeyFromSeed(seedK), audience, req, sig) {
		t.Error("verified under the wrong key")
	}
	msg, _ := possession.MessageBytes(req.Nonce, binding)
	if Verify(pubP, audience, req, crypto.Sign(seedP, msg)) {
		t.Error("a raw signature verified as a login proof")
	}
	other, _ := possession.Prove(seedP, "archon-login/2", req.Nonce, binding)
	if Verify(pubP, audience, req, other) {
		t.Error("a proof from another domain verified")
	}
	collect, _ := ProveCollect(seedK, audience, req)
	if Verify(crypto.PublicKeyFromSeed(seedK), audience, req, collect) {
		t.Error("a collect proof verified as a login proof")
	}
}

func TestCollect(t *testing.T) {
	req := request()
	sig, err := ProveCollect(seedK, audience, req)
	if err != nil {
		t.Fatal(err)
	}
	if !VerifyCollect(audience, req, sig) {
		t.Fatal("collect proof did not verify")
	}
	if VerifyCollect("https://evil.example/api", req, sig) {
		t.Error("collect proof verified at another audience")
	}
	login, _ := Prove(seedK, audience, req)
	if VerifyCollect(audience, req, login) {
		t.Error("a login proof verified as a collect proof")
	}
	if _, err := ProveCollect(seedP, audience, req); err == nil {
		t.Error("collect proof from a key the request does not name was not refused")
	}
}

func TestRefusals(t *testing.T) {
	cases := []struct {
		name string
		aud  string
		mod  func(*Request)
	}{
		{"empty audience", "", func(*Request) {}},
		{"control char in audience", "https://dawn.example/api\n", func(*Request) {}},
		{"browser wrong size", audience, func(r *Request) { r.Browser = r.Browser[:31] }},
		{"empty id", audience, func(r *Request) { r.ID = nil }},
		{"empty scope entry", audience, func(r *Request) { r.Scope = []string{""} }},
		{"control char in scope", audience, func(r *Request) { r.Scope = []string{"read:\x07projects"} }},
		{"DEL in scope", audience, func(r *Request) { r.Scope = []string{"read:\x7fprojects"} }},
		{"invalid utf-8 in scope", audience, func(r *Request) { r.Scope = []string{"read:\xffprojects"} }},
		{"zero validity", audience, func(r *Request) { r.ValidFor = 0 }},
		{"oversized id", audience, func(r *Request) { r.ID = bytes.Repeat([]byte{1}, MaxFieldSize+1) }},
		{"binding over the possession bound", audience, func(r *Request) {
			r.Scope = []string{strings.Repeat("a", MaxFieldSize), strings.Repeat("b", MaxFieldSize)}
		}},
	}
	for _, tc := range cases {
		r := request()
		tc.mod(r)
		if _, err := Binding(RoleLogin, tc.aud, r); err == nil {
			t.Errorf("%s: not refused", tc.name)
		}
		if _, err := Prove(seedP, tc.aud, r); err == nil {
			t.Errorf("%s: Prove did not refuse", tc.name)
		}
		if Verify(crypto.PublicKeyFromSeed(seedP), tc.aud, r, make([]byte, 64)) {
			t.Errorf("%s: Verify returned true", tc.name)
		}
	}
	if _, err := Binding(0x03, audience, request()); err == nil {
		t.Error("unknown role not refused")
	}
	r := request()
	r.Nonce = r.Nonce[:15]
	if _, err := Prove(seedP, audience, r); err == nil {
		t.Error("short nonce not refused by Prove")
	}
	r = request()
	r.Scope = nil
	if _, err := Binding(RoleLogin, audience, r); err != nil {
		t.Errorf("empty scope list should be allowed (the law may refuse it): %v", err)
	}
}

func TestDeterministic(t *testing.T) {
	a, _ := Prove(seedP, audience, request())
	b, _ := Prove(seedP, audience, request())
	if hex.EncodeToString(a) != hex.EncodeToString(b) {
		t.Fatal("login proof is not deterministic")
	}
}
